### PR TITLE
workflows/scheduled-snapshots: do not update relased fedora repos

### PR DESCRIPTION
**Pull request by jrusz**
Fedora and fedora-modular repositories don't change so don't update them
when opening a PR with snapshot update.

I also had to switch to perl to be able to use the negative lookbehind. I don't know any better solution this is the first one I've put together with a lot of trial and error and googling. Suggestions more than welcome. This change is needed once https://github.com/osbuild/osbuild-composer/pull/2508 lands to update the snapshots correctly.

### CODE REVIEW

**Comment by jrusz** [comment]
Also @Gundersanne can you please change the permissions here so that a review is required before merge? The green button looks scary when it's not backed by a review :sweat_smile:

File: .github/workflows/scheduled-snapshots.yml
Comment (Gundersanne) [inline .github/workflows/scheduled-snapshots.yml:111]: I don't fully understand the reason, why would we ever specify fedora repositories?

Comment (jrusz) [reply .github/workflows/scheduled-snapshots.yml:111]: I am proposing a change where even fedora repositories will be pinned in Schutzfile and since `fedora` and `fedora-modular` repository snapshots don't get updated (because they don't change) it will get broken if it's changed by the PR that this action creates.

**Comment by gicmo** [comment]
> 

I took the liberty to do that.

**Comment by dvdhrm** [comment]
These repos are marked as singletons in their json config. This causes them to be skipped if they exist.

**Comment by jrusz** [comment]
Ugh, ok so I may have found a way how to edit the correct values with `jq` but I don't know how do I make it into a json again :/ 
```
 jq --arg SUFFIX "$SUFFIX" '.[] | select(has("repos")) | .repos[] | .x86_64[], .aarch64[] | .baseurl | if . | contains("-fedora-") or contains("-fedora-modular-") then . else sub("[0-9]{8}";"$SUFFIX") end' Schutzfile
```
That only outputs all the urls with the changed date

**Comment by jrusz** [comment]
Ok I think I got it.

File: .github/workflows/scheduled-snapshots.yml
Comment (ondrejbudai) [inline .github/workflows/scheduled-snapshots.yml:113]: This is still quite crazy, can you comment in the file what it does so our future selves know what it does?

File: .github/workflows/scheduled-snapshots.yml
Comment (ondrejbudai) [inline .github/workflows/scheduled-snapshots.yml:150]: I think that this script is the same one as above. What about extracting it to a new file and reusing it? Would it be viable?

**Review by ondrejbudai — CHANGES_REQUESTED** [review]
Very nice, I think it just needs a bit of a clean-up and more comments. ;)

File: .github/workflows/scheduled-snapshots.yml
Comment (jrusz) [reply .github/workflows/scheduled-snapshots.yml:113]: I suppose I could run `grep -r "singleton"` in the repos directory and get a list of repository names from that and use it somehow? I'll try something like that.

Comment (jrusz) [reply .github/workflows/scheduled-snapshots.yml:113]: > We could put something in the json itself indicating to skip/update the baseurl in the workflow, or we could require the baseurl to be an rpmrepo one (whitelist rather than blacklist).

That's a great suggestion, thanks @Gundersanne I didn't realize we can put whatever in the `Schutzfile` so I'll just add a boolean to each repo file to indicate whether it can be updated or not which will also greatly simplify the `jq` command.

Comment (jrusz) [reply .github/workflows/scheduled-snapshots.yml:113]: So I've created https://github.com/osbuild/osbuild-composer/pull/2590 which will simplify the `jq` expression to:
```
(.[] | select(has("repos")) | .repos[] | select(.updateable) | .x86_64[], .aarch64[] | .baseurl) |= sub("[0-9]{8}"; "SUFFIX")
```
This should make more sense and also makes it easy to just affect selected repositories.

File: .github/workflows/scheduled-snapshots.yml
Comment (jrusz) [reply .github/workflows/scheduled-snapshots.yml]: > I think that this script is the same one as above. What about extracting it to a new file and reusing it? Would it be viable?

Yes, thanks for the suggestion, I've put the whole thing inside a script.

**Comment by jrusz** [comment]
This needs the two other related PRs to get merged first.

**Comment by jrusz** [comment]
@ondrejbudai I came up with "something". Can you take a look and share what you think about this? I've put the whole updating logic in a script and the way it works now is get's all the `baseurl` values from Schuztfile and iterates over them. In the loop it finds the corresponding definition from this repository and checks for the presence of `singleton` key. If it's present the url is skipped and not updated, otherwise it's updated. Once that's done the PR is opened like before.

File: .github/scripts/snapshot_update_pr.sh
Comment (ondrejbudai) [inline .github/scripts/snapshot_update_pr.sh:7]: I just wonder if lines 7-33 aren't worth a small python script. I completely get that `jq` is powerful but the readability is just hard. I can imagine that this would be slightly more comprehensive when written in a more imperative style (e.g. line 28).

What are your thougths?

Comment (jrusz) [reply .github/scripts/snapshot_update_pr.sh:7]: My thoughts are that I have no idea how this would look in Python but you're probably right and I'll try it to see how it will look like :smile:

Comment (jrusz) [reply .github/scripts/snapshot_update_pr.sh]: Well I again created "something"...  Tbh I don't think it's much simpler, but maybe I just suck at python :D looking forward to another review. I kept both commits just for comparison sake, I'll squash them before merge :)

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:13]: Extra parentheses.

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:7]: Python prefers lowercase letters for variable names.

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:16]: I suggest passing this as an argument.

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:25]: Parentheses once again. :)

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:32]: This can be probably removed.

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:30]: `else` is not needed here.

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:26]: I think we can even do `for repo in data[key]['repos']` here.

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:22]: This can be an argument too.

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:17]: `os.path.join` once again. :)

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:34]: `os.path.join`

**Review by ondrejbudai — CHANGES_REQUESTED** [review]
I like the direction of this, the python script is way more readable for me that the `jq` magic. I put some suggestions inline.

File: .github/scripts/update_schutzfile.py
Comment (jrusz) [reply .github/scripts/update_schutzfile.py]: Yes, I probably had it there for debugging.

File: .github/scripts/update_schutzfile.py
Comment (jrusz) [reply .github/scripts/update_schutzfile.py]: Actually not cause I need numbers to index the list in the dict. Maybe if there's a simpler way to get the length?

File: .github/scripts/update_schutzfile.py
Comment (jrusz) [reply .github/scripts/update_schutzfile.py]: > Also, I'm not certain if the path should be absolute, it feels wrong, does it actually work?

I guess not, that looks like an oversight when converting from my debug version to the final version here :)

File: .github/scripts/update_schutzfile.py
Comment (dvdhrm) [reply .github/scripts/update_schutzfile.py]: `for key, value in enumerate(data[key]['repos'])`

Comment (jrusz) [reply .github/scripts/update_schutzfile.py]: > `for key, value in enumerate(data[key]['repos'])`

Perfect, thanks!

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:16]: I think there was a misunderstanding in the last review. I think `REPO` and `SUFFIX` shouldn't be passed inside the script using environmental variables but using command line arguments. A simple `argparse` should be enough. See e.g. https://github.com/osbuild/maintainer-tools/blob/main/update-distgit.py#L103

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:24]: This can indeed be `for repo in data[key]["repos"]` because `repo` is a dict and thus it will be a reference (not a copy).

See https://www.online-python.com/jr9hdI0fFw

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:25]: If the previous suggestion is applied, we can make this `for arch_repos in repo` and just use `arch_repo` onwards, e.g.:

```
if any(singleton in arch_repos[0]["baseurl"] for singleton in singletons):
```

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:20]: I would drop this so this can work for any architecture, see my comment below.

File: .github/scripts/update_schutzfile.py
Comment (jrusz) [reply .github/scripts/update_schutzfile.py]: Ah sorry, yeah I didn't understand it correctly. I think I got it now :)

File: .github/scripts/update_schutzfile.py
Comment (jrusz) [reply .github/scripts/update_schutzfile.py]: I got a bit stuck on this cause `repo` in this case is still a dict which has a string `file` and then two lists for each arch, but was able to do it without the enumeration in the end.

File: .github/scripts/update_schutzfile.py
Comment (jrusz) [reply .github/scripts/update_schutzfile.py]: Yeah ok got it working with some adjustments of the following code

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:1]: We usually use

```suggestion
#!/usr/bin/python3
```

File: .github/scripts/update_schutzfile.py
Comment (ondrejbudai) [inline .github/scripts/update_schutzfile.py:48]: We need to iterate over all repos here, otherwise it will update only the first one. (I tested the script locally on composer's Schutzfile and it updated only the first repository for a specific distroarch combination (only for CentOS and RHEL)).

File: .github/workflows/scheduled-snapshots.yml
Comment (ondrejbudai) [inline .github/workflows/scheduled-snapshots.yml:131]: You need to pass the parameters here (I believe you can even pass `--suffix ${{ steps.generate_suffix.outputs.suffix }}` directly without an env variable).

**Review by ondrejbudai — CHANGES_REQUESTED** [review]
Thanks for bearing with me! We are getting near, though! Stuff inline:

- 2 stylistic things
- a bug preventing some repos from being updated
- forgotten arguments in the github action

After this is fixed, I believe that this is perfect and it can go in!

File: .github/scripts/update_schutzfile.py
Comment (jrusz) [reply .github/scripts/update_schutzfile.py:48]: Aaah, damn, yeah you're right, thanks for noticing, I've checked the fedoras and then saw a lot of changes so I thought it was fine :disappointed:

File: .github/workflows/scheduled-snapshots.yml
Comment (jrusz) [reply .github/workflows/scheduled-snapshots.yml:131]: Do I? This is the bash script which then calls the python script with the parameters... And in that case I guess I want to keep the env vars... Or you think I should first run the python script and then the bash script in the workflow here?

**Comment by jrusz** [comment]
> Thanks for bearing with me! We are getting near, though! Stuff inline:
> 
> * 2 stylistic things
> * a bug preventing some repos from being updated
> * forgotten arguments in the github action
> 
> After this is fixed, I believe that this is perfect and it can go in!

Well I thank you for all these reviews! :) ... I am unsure about one requested change though. Currently I am still calling the bash script from the workflow which then calls the python script with the parameters. The reason being that the bash scrip then runs the github commands to open the PR. I don't know if this structure is fine or if it would be better to first call the python script which will just prepare the Schutzfile and then call the bash script which will do the rest... I guess the second option of running both scripts from the workflow one after another seems better/more clear?

File: .github/workflows/scheduled-snapshots.yml
Comment (ondrejbudai) [reply .github/workflows/scheduled-snapshots.yml:131]: I got confused, you are right. :)

**Review by ondrejbudai — APPROVED** [review]

**Pull request merged**
